### ISSUE TITLE
start_decay_score_delta option not being applied

### ISSUE THREAD

**Issue by shahbazsyed** [OpenNMT/OpenNMT#330]
Hi,
My training configuration is as follows 
`layers=3
max_batch_size=16
encoder_type=brnn
rnn_size=600
learning_rate_decay=0.5
start_decay_score_delta=5
decay=score_only
optim=sgd
learning_rate=1
dropout=0.2
end_epoch=25`

However, I observed that the learning rate is not being decreased even after the difference between perplexity scores from epoch to epoch is less than 5. 

```
[07/02/17 02:52:48 INFO] Epoch 14 ; Iteration 6350/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1566 ; Perplexity 53.26
[07/02/17 02:55:09 INFO] Epoch 14 ; Iteration 6400/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1550 ; Perplexity 53.95
[07/02/17 02:57:23 INFO] Epoch 14 ; Iteration 6450/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1564 ; Perplexity 51.31
[07/02/17 02:59:42 INFO] Epoch 14 ; Iteration 6500/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1444 ; Perplexity 50.32
[07/02/17 03:01:57 INFO] Epoch 14 ; Iteration 6550/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1587 ; Perplexity 52.29
[07/02/17 03:04:27 INFO] Epoch 14 ; Iteration 6600/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1504 ; Perplexity 54.91
[07/02/17 03:06:40 INFO] Epoch 14 ; Iteration 6650/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1543 ; Perplexity 51.89
[07/02/17 03:09:04 INFO] Epoch 14 ; Iteration 6700/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1580 ; Perplexity 53.73
[07/02/17 03:11:29 INFO] Epoch 14 ; Iteration 6750/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1554 ; Perplexity 52.29
[07/02/17 03:13:42 INFO] Epoch 14 ; Iteration 6800/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1574 ; Perplexity 53.63
[07/02/17 03:15:58 INFO] Epoch 14 ; Iteration 6850/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1563 ; Perplexity 52.97
[07/02/17 03:18:15 INFO] Epoch 14 ; Iteration 6900/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1520 ; Perplexity 54.13
[07/02/17 03:20:37 INFO] Epoch 14 ; Iteration 6950/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1591 ; Perplexity 52.94
[07/02/17 03:22:51 INFO] Epoch 14 ; Iteration 7000/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1521 ; Perplexity 54.29
[07/02/17 03:23:24 INFO] Epoch 14 ; Iteration 7014/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1518 ; Perplexity 52.46
[07/02/17 03:23:24 INFO] Evaluating on the validation dataset...
[07/02/17 03:32:07 INFO] Validation perplexity: 80.13
[07/02/17 03:32:07 INFO] Saving checkpoint to '/root/deeplearning/vulgar/stopwords/vulgar_epoch14_80.13.t7'...
[07/02/17 03:32:11 INFO]
[07/02/17 03:34:30 INFO] Epoch 15 ; Iteration 50/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1539 ; Perplexity 46.93
[07/02/17 03:36:34 INFO] Epoch 15 ; Iteration 100/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1573 ; Perplexity 44.65
[07/02/17 03:38:47 INFO] Epoch 15 ; Iteration 150/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1571 ; Perplexity 46.51
[07/02/17 03:41:11 INFO] Epoch 15 ; Iteration 200/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1470 ; Perplexity 45.59
[07/02/17 03:43:24 INFO] Epoch 15 ; Iteration 250/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1626 ; Perplexity 44.44
```

```
[07/02/17 08:10:58 INFO] Epoch 15 ; Iteration 6150/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1603 ; Perplexity 48.50
[07/02/17 08:13:21 INFO] Epoch 15 ; Iteration 6200/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1430 ; Perplexity 50.40
[07/02/17 08:15:46 INFO] Epoch 15 ; Iteration 6250/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1590 ; Perplexity 49.18
[07/02/17 08:18:07 INFO] Epoch 15 ; Iteration 6300/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1604 ; Perplexity 49.16
[07/02/17 08:20:29 INFO] Epoch 15 ; Iteration 6350/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1576 ; Perplexity 51.36
[07/02/17 08:22:48 INFO] Epoch 15 ; Iteration 6400/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1491 ; Perplexity 48.49
[07/02/17 08:25:13 INFO] Epoch 15 ; Iteration 6450/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1509 ; Perplexity 49.62
[07/02/17 08:27:21 INFO] Epoch 15 ; Iteration 6500/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1510 ; Perplexity 50.92
[07/02/17 08:29:32 INFO] Epoch 15 ; Iteration 6550/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1577 ; Perplexity 51.78
[07/02/17 08:31:49 INFO] Epoch 15 ; Iteration 6600/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1504 ; Perplexity 51.90
[07/02/17 08:34:18 INFO] Epoch 15 ; Iteration 6650/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1496 ; Perplexity 52.86
[07/02/17 08:36:37 INFO] Epoch 15 ; Iteration 6700/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1548 ; Perplexity 51.59
[07/02/17 08:38:48 INFO] Epoch 15 ; Iteration 6750/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1565 ; Perplexity 51.10
[07/02/17 08:41:02 INFO] Epoch 15 ; Iteration 6800/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1495 ; Perplexity 52.08
[07/02/17 08:43:23 INFO] Epoch 15 ; Iteration 6850/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1483 ; Perplexity 51.77
[07/02/17 08:45:46 INFO] Epoch 15 ; Iteration 6900/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1505 ; Perplexity 51.45
[07/02/17 08:48:15 INFO] Epoch 15 ; Iteration 6950/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1582 ; Perplexity 48.18
[07/02/17 08:50:38 INFO] Epoch 15 ; Iteration 7000/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1524 ; Perplexity 48.22
[07/02/17 08:51:12 INFO] Epoch 15 ; Iteration 7014/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1679 ; Perplexity 47.46
[07/02/17 08:51:12 INFO] Evaluating on the validation dataset...
[07/02/17 09:00:17 INFO] Validation perplexity: 82.11
[07/02/17 09:00:17 INFO] Saving checkpoint to '/root/deeplearning/vulgar/stopwords/vulgar_epoch15_82.11.t7'...
[07/02/17 09:00:21 INFO]
[07/02/17 09:02:39 INFO] Epoch 16 ; Iteration 50/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1581 ; Perplexity 42.38
[07/02/17 09:04:58 INFO] Epoch 16 ; Iteration 100/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1545 ; Perplexity 43.77
[07/02/17 09:07:10 INFO] Epoch 16 ; Iteration 150/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1646 ; Perplexity 43.44
[07/02/17 09:09:19 INFO] Epoch 16 ; Iteration 200/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1543 ; Perplexity 44.95
[07/02/17 09:11:45 INFO] Epoch 16 ; Iteration 250/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1490 ; Perplexity 44.67
[07/02/17 09:14:05 INFO] Epoch 16 ; Iteration 300/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1512 ; Perplexity 44.68
[07/02/17 09:16:17 INFO] Epoch 16 ; Iteration 350/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1603 ; Perplexity 44.39
[07/02/17 09:18:40 INFO] Epoch 16 ; Iteration 400/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1529 ; Perplexity 45.29
[07/02/17 09:20:53 INFO] Epoch 16 ; Iteration 450/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1680 ; Perplexity 40.58
[07/02/17 09:23:01 INFO] Epoch 16 ; Iteration 500/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1528 ; Perplexity 44.88
[07/02/17 09:25:19 INFO] Epoch 16 ; Iteration 550/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1544 ; Perplexity 43.39
[07/02/17 09:27:45 INFO] Epoch 16 ; Iteration 600/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1414 ; Perplexity 43.81
[07/02/17 09:30:09 INFO] Epoch 16 ; Iteration 650/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1518 ; Perplexity 44.60
[07/02/17 09:32:22 INFO] Epoch 16 ; Iteration 700/7014 ; Optim SGD LR 1.0000 ; Source tokens/s 1615 ; Perplexity 43.50
```

Is there something else to be changed in the configuration?

**Comment by guillaumekln** [comment]
Thank for reporting! It was an actual bug as the delta was applied in the incorrect direction.

The code now reflects what is described in the documentation, namely the learning rate is decayed if the validation score is not improving more than `start_decay_score_delta`.
